### PR TITLE
[#804] OPTIONS request mocking support

### PR DESCRIPTION
**Pull request by riklaunim**
Initial commit for OPTIONS request mocking. Does not work yet, throwing:

```
TypeError: this.pretender[verb] is not a function
```

When options is trying to get mocked.

### CODE REVIEW

File: addon/route-handler.js [hunk 1]
```diff
@@ -41,6 +41,8 @@ function createHandler({ verb, schema, serializerOrRegistry, path, rawHandler, o
     handler = new DeleteShorthandHandler(...args);
   } else if (verb === 'head') {
     handler = new HeadShorthandHandler(...args);
+  } else if (verb === 'options') {
+    handler = new GetShorthandHandler(...args);
```
Comment (riklaunim) [inline addon/route-handler.js]: This will rather require it's own handler.

**Comment by samselikoff** [comment]
Looks like [Pretender supports OPTIONS](https://github.com/pretenderjs/pretender/blob/7e2d2dcef8fe6fe0dc00807a8da2ca337663a06d/test/registry_test.js#L15).

Regarding the shorthand handler, I agree it should have its own handler. Should there even be a shorthand for OPTIONS? What would the behavior be?

Perhaps we can start with just adding the verb, and letting users use the function handler, e.g.

```js
this.options('/some/url', () => {
  // normal function handler response
});
```

**Comment by ksnyde** [comment]
@samselikoff I developed a PR but it looks too similar to this one to add it to the repo. the sad part though is that neither seems to address allowing a passthrough of a OPTIONS & POST pairing.

**Comment by ksnyde** [comment]
Ok, my last comment is a bit misleading. I had in earlier testing removed the "options" from the my `config.js`. Now that it is correct to read:

    this.passthrough('https://www.googleapis.com/identitytoolkit/v3/relyingparty/verifyPassword', ['post', 'head', 'options']);

I'm back to my state from yesterday where it is fails on the "options" verb of the pretender object: 
https://github.com/samselikoff/ember-cli-mirage/blob/master/addon/server.js#L179
image

**Comment by ksnyde** [comment]
Sorry to be so verbose but ... another update ...

I switched the URL to a shortened version with wildcard characters:

    this.passthrough('https://www.googleapis.com/**', ['post', 'head', 'options']);

this DOES work with this PR. 👍

**Comment by ksnyde** [comment]
Any chance we could merge this in? even with the fully qualified URL it is an improvement I think.

**Comment by riklaunim** [comment]
I need to fix the style validation errors and see what else is failing. Either way you can use a git commit as the version in package.json if you just want to use it now :)

```
git+https://github.com/riklaunim/ember-cli-mirage.git#abae084fe2046648a7b26b31e49df2841c9f90fb
```
should do the trick.

**Comment by ksnyde** [comment]
yup @riklaunim that's what I'm doing for now. glad you got to this 18 days before me. :) Thanks for the PR.

One last note, working off your code base I noticed that even though my shortened URL works the passthrough call will fail when explicitly asking for 'options'.

**Comment by ksnyde** [comment]
@riklaunim in fact I _can't_ use yours yet as it fails on the same line as it was with base ember-cli-mirage. I had added one piece of defensive code into my clone of your PR:

    verbs.forEach(verb => {
      paths.forEach(path => {
        let fullPath = this._getFullPath(path);
        if(this.pretender[verb]) {
          this.pretender[verb](fullPath, this.pretender.passthrough);
        } else {
          console.warn(`Call to a non-existant pretender method: "${verb}". Pretender is: `, this.pretender);
        }
      });
    });

Not sure whether this is worth having in the final code once everything is sorted but it does get this to working.

**Comment by riklaunim** [comment]
Well, I was wondering why Pretender doesn't have my options and other changes visible. Turns out the installed dependency isn't the one that is being used (that's from a docker container):

```
./bower_components/pretender/pretender.js
./node_modules/ember-cli-mirage/node_modules/pretender/pretender.js
```
node_modules/ember-cli-mirage/node_modules/pretender/pretender.js is used and not the version I install with my project as ./bower_components/pretender/pretender.js

And Pretender seems to need this https://github.com/pretenderjs/pretender/pull/166 to work. When I added that to the used pretender.js file - options do mock:

```
this.options('/campaigns', function () {
      return .....;
});
```

**Comment by ksnyde** [comment]
@riklaunim got a bit confused on your npm versus bower **Pretender** versions comment. I'd expect that Mirage is only using the npm version but I think that's what you're saying. Anyway, glad to see your PR on Pretender as I pretty sure it had to originate from there based on what we were seeing hanging off the Pretender object. Sounds like you've nailed that.

**Comment by ksnyde** [comment]
One other dimension to this that I've now noticed, is that as soon as I add the _passthrough_ to my `config.js` this somehow interferes with Mirage using my `scenarios/default.js` which is loading my fixtures and factories. It's very reproducible but haven't had time to dig into one might effect the other.

@samselikoff does anything jump out at you WRT to this?

**Comment by riklaunim** [comment]
installing ember-cli-mirate adds pretender to bower dependencies so I expected it will use it from bower_components or if its want it in node_modules then it will be in node_modules and not /node_modules/subfolder/node_modules/ -- which is pretty hidden.

**Comment by ksnyde** [comment]
Oh yes that is confusing. @samselikoff is there a reason for this?

**Comment by ksnyde** [comment]
btw, removing my above comment until I can dig into it a bit further ... it's not as reproducible as i thought 😬

**Comment by ksnyde** [comment]
@riklaunim I have forked your PR and pointed to your Pretender PR in the `package.json` so I get the full end-to-end effect. Works. I can send you the PR if you want but I'm guessing you'd prefer to keep these two PR separate even though the Mirage PR depends on the Pretender one.

**Comment by ksnyde** [comment]
I can confirm though that a side effect does seem to be introduced. When I use this PR paired with the Pretender PR, I get no errors but Mirage _does not_ use my `config/default.js` configuration and therefore fixtures and factories are not used.

**Use Case 1**: PR breaks scenarios
**Use Case 2**: PR solves options problem

So to be clear, use case 1a:
- using published Mirage 0.2.1
- remove all passthrough configuration in `mirage/config.js` ... without which no OPTIONS calls will be made
- load an application page who's model ('forecast') has an associated fixture associated with it
- the `GET /forecasts` request results in the expected 80 rows that come from the fixture
- no error, expected outcome

Use case 1b:
- Use this PR alongside the Pretender PR
- exact same setup as 1a, outcome though is different ... no error, just no results


Use Case 2a:
- using default Mirage 0.2.1
- go to a page where `POST https://www.googleapis.com` is called (and indirectly the OPTIONS call)
- add passthrough --`this.passthrough('https://www.googleapis.com/**', ['post', 'head', 'options']);` -- to `mirage/config.js`.
- immediately fails on any page with the above mentioned "_this.pretender[verb] is not a function" error

Use Case 2b: 
- using Mirage and Pretender PR's (see [link](https://github.com/riklaunim/ember-cli-mirage/compare/master...ksnyde:master))
- No error, network request goes out as expected ... desired behaviour.

**Comment by riklaunim** [comment]
Pretender needs to merge the PR and push a new version, then this can use it and not a github commit (which would be pretty odd for a production release).

**Comment by ksnyde** [comment]
yes that was my point 👍 

I just did the fork for my immediate needs of seeing the two together.

**Comment by ksnyde** [comment]
the issue above _does_ need addressing though as Mirage will not work correctly without addressing this

**Comment by riklaunim** [comment]
I use it with a big client panel demo and it works. If the are problems - test should show them too - when the verb problem is solved by newer pretender.

**Comment by ksnyde** [comment]
so you're loading fixtures and/or factories to initialise your state?

**Comment by riklaunim** [comment]
Mostly factories, few custom responses. I have some passthrough in another project, can check it too.

**Comment by ksnyde** [comment]
would you let me know how that comes out? I have to shift to a different project for a day but if you're not seeing this behaviour I'll dig into mine again and see what else could be causing it.

**Comment by riklaunim** [comment]
Tested, works.

File: addon/route-handler.js (hunk 1 above)
Comment (samselikoff) [inline addon/route-handler.js]: can we make this the FunctionHandler? I don't think shorthands make sense for options (or do they?)

**Comment by samselikoff** [comment]
I'd love the tl;dr from your convo & experiments - what should our next steps be here? Are we blocked by the Pretender pr?

**Comment by riklaunim** [comment]
At least some tests won't pass as Pretender.options does not exist. Then there may be some other options tests for fixing. It's also odd that node_modules/ember-cli-migrage/node_modules/.... subfolders shows up with Pretender - and not just the one in bower_components. I had to hand modify the Pretender code there instead of getting it from a git version in bower dependencies.

I'll see about the Pretender additional tests, then see those tests here.

**Comment by samselikoff** [comment]
Awesome, thank you. So we should hold off on this PR until those are sorted out, right?

**Comment by riklaunim** [comment]
Now fixture tests fail for some reason.

File: package.json [hunk 2]
```diff
@@ -75,7 +75,7 @@
     "exists-sync": "0.0.3",
     "fake-xml-http-request": "^1.4.0",
     "faker": "3.0.1",
-    "pretender": "^1.2.0",
+    "pretender": "^1.3.0",
```
Comment (samselikoff) [inline package.json]: I think we can remove this change from this PR now? Once you do the tests will rerun and we can investigate any errors.

Let's get this puppy merged!

Comment (riklaunim) [inline package.json]: 1.2.0 doesn't have OPTIONS support. 1.3.0 is the first one to have it.

Comment (samselikoff) [inline package.json]: right but `master` is now on 1.4 so if you remove this & rebase we should be set

Comment (riklaunim) [inline package.json]: ah, rebase to master. OK :)

**Comment by samselikoff** [comment]
You have some failures, are your tests passing locally?

**Comment by riklaunim** [comment]
nop, as I commented above - few fixture related tests fail and I don't why.

**Comment by riklaunim** [comment]
Can you help me with this? It seems like even adding options to the list - https://github.com/samselikoff/ember-cli-mirage/pull/841/commits/14523a7d52163e9f6dd44acb4fd989f766e40bd6#diff-3da129bcd0b0367d1b65b28e9ecc34c9 - causes the fixtures tests to fail. For some reason visit() return 404 - like if some mock or something stopped working.

**Comment by samselikoff** [comment]
Closing due to inactivity

**Pull request closed without merging**